### PR TITLE
Wasserstein

### PR DESCRIPTION
**Pull request by GaganCodes**
Please read through our [contribution guide](https://github.com/pytorch-labs/torcheval/blob/main/CONTRIBUTING.md) prior to creating your pull request.

Summary:
Implemented the wasserstein distance function and metric class for 1 dimensional case.

Test plan:
Every feature of the code is tested via unitTest framework in "torcheval\tests\metrics\statistical\test_wasserstein.py" and "torcheval\tests\metrics\functional\statistical\test_wasserstein.py".

Fixes #{issue number}
https://github.com/pytorch/torcheval/issues/137

### CODE REVIEW

File: tests/metrics/statistical/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
```
Comment (bobakfb) [inline tests/metrics/statistical/__init__.py:5]: Please add the wasserstein metric here inside the `__all__` list.

**Review by bobakfb — CHANGES_REQUESTED** [review]
Hey @GaganCodes, thanks for this, looks great!

Could you remove the few extra files you added here: `.vscode/...` and the sphinx theme submodule?

**Comment by GaganCodes** [comment]
Thank you for the quick feedback @bobakfb, I'll fix these and get back to you ASAP.

**Comment by GaganCodes** [comment]
@bobakfb code lint and extra files comments are resolved. Thanks!

File: tests/metrics/functional/statistical/test_wasserstein.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                    my_compute_result_cuda,
+                    compute_result_cuda,
+                    equal_nan=True,
+                    atol=1e-8,
+                    rtol=1e-5,
+                )
+
+    def test_wasserstein1d_distribution_values_only(self) -> None:
+        x = torch.Tensor([5, -5, -7, 9, -3])
+        y = torch.Tensor([9, -7, 5, -4, -2])
```
Comment (ananthsub) [inline tests/metrics/functional/statistical/test_wasserstein.py:117]: here and everywhere else in the tests: use `torch.tensor` for the constructor
```suggestion
        x = torch.tensor([5, -5, -7, 9, -3])
        y = torch.tensor([9, -7, 5, -4, -2])
```

File: tests/metrics/statistical/test_wasserstein.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            x_weights_np = x_weights.numpy().flatten()
+        if y_weights is not None:
+            y_weights_np = y_weights.numpy().flatten()
+
+        scipy_result = np.stack(
+            [
+                sp_wasserstein(sp_x, sp_y, sp_x_w, sp_y_w)
+                for sp_x, sp_y, sp_x_w, sp_y_w in zip(
+                    [x_np], [y_np], [x_weights_np], [y_weights_np]
+                )
```
Comment (bobakfb) [inline tests/metrics/statistical/test_wasserstein.py:47]: same as above, I think this is equivelant to 

```
[sp_wasserstein(x_np, y_np, x_weights_np, y_weights_np),]
```

File: torcheval/metrics/functional/statistical/wasserstein.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        y_cdf = y_cdf_indices.to(device) / y.size(0)
+    else:
+        y_sorted_cum_weights = torch.cat(
+            (torch.Tensor([0]).to(device), torch.cumsum(y_weights[y_sorter], dim=0))
+        )
+        y_cdf = y_sorted_cum_weights[y_cdf_indices] / y_sorted_cum_weights[-1]
+
+    return torch.sum(
+        torch.multiply(torch.abs(x_cdf - y_cdf), deltas), dim=0, keepdim=True
+    ).to(device)
```
Comment (bobakfb) [inline torcheval/metrics/functional/statistical/wasserstein.py:185]: This is awesome! Super high quality!!

File: torcheval/metrics/functional/statistical/wasserstein.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    >>> wasserstein_1d(torch.tensor([0,1,2,2]), torch.tensor([0,1]))
+    torch.tensor([0.75])
+
+    """
+    _wasserstein_param_check(x, y, x_weights, y_weights)
+    _wasserstein_update_input_check(x, y, x_weights, y_weights)
+    return _wasserstein_compute(x, y, x_weights, y_weights)
+
+
+def _wasserstein_param_check(
```
Comment (bobakfb) [inline torcheval/metrics/functional/statistical/wasserstein.py:92]: small refactor here, `params` for us are things that don't change between calls to `update`, e.g. constructor args. It seems the checks in `param_check` should be in `update_input_check`. It looks like you don't have any params, so you can just delete this function and add the checks to `update_input_check`.

**Review by bobakfb — CHANGES_REQUESTED** [review]
Hey overall this is incredible! Thanks so much for the contribution. There are a few style pointers

1. use `torch.tensor()` instead of constructing the object directly with `torch.Tensor`
2. remove `param_input_check` and move checks into `update_input_check`
3. You have a bunch of unused imports, can you double check the lint output for them? I can remove the unused imports for you if this is hard.

File: torcheval/metrics/functional/statistical/wasserstein.py (hunk 5 above)
Comment (GaganCodes) [reply torcheval/metrics/functional/statistical/wasserstein.py:92]: Thank you for this thorough review @bobakfb ! I will try to get this done over the weekend.

File: tests/metrics/functional/statistical/test_wasserstein.py (hunk 2 above)
Comment (GaganCodes) [reply tests/metrics/functional/statistical/test_wasserstein.py]: These changes have been implemented. Please help me check them, thanks!

File: tests/metrics/functional/statistical/test_wasserstein.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            x_weights_np = x_weights.numpy()
+        if y_weights is not None:
+            y_weights_np = y_weights.numpy()
+
+        if x.ndim == 1:
+            scipy_result = np.stack(
+                [
+                    sp_wasserstein(sp_x, sp_y, sp_x_w, sp_y_w)
+                    for sp_x, sp_y, sp_x_w, sp_y_w in zip(
+                        [x_np], [y_np], [x_weights_np], [y_weights_np]
```
[the remark this answers is not in the document]
Comment (GaganCodes) [reply tests/metrics/functional/statistical/test_wasserstein.py]: This change has been implemented, thanks!

File: tests/metrics/statistical/test_wasserstein.py (hunk 3 above)
Comment (GaganCodes) [reply tests/metrics/statistical/test_wasserstein.py]: This change has been implemented, thanks!

File: torcheval/metrics/functional/statistical/wasserstein.py (hunk 4 above)
Comment (GaganCodes) [reply torcheval/metrics/functional/statistical/wasserstein.py:176]: Thank you for the kind words!

File: torcheval/metrics/functional/statistical/wasserstein.py (hunk 5 above)
Comment (GaganCodes) [reply torcheval/metrics/functional/statistical/wasserstein.py]: This change has been implemented, thanks!

**Comment by GaganCodes** [comment]
Hi @bobakfb , @ananthsub , I am not sure why pre-commit is failing for flake8, it's showing "Passed" on my local machine. I have also manually ensured that no imported module is unused.

For transparency, python version on my machine is 3.10.12, whereas in pre-commit check here it seems to be
